manul/src/combinators/extend.rs

```diff
@@ -0,0 +1,378 @@
+    fn make_normal_broadcast(
+        &self,
+        rng: &mut impl CryptoRngCore,
+        round: &Self::Round,
+    ) -> Result<<Self::Round as Round<Id>>::NormalBroadcast, LocalError> {
+        round.make_normal_broadcast(rng)
+    }
+
+    /// Called instead of [`Round::make_echo_broadcast`].
+    ///
+    /// The default implementation calls [`Round::make_echo_broadcast`].
+    fn make_echo_broadcast(
+        &self,
+        rng: &mut impl CryptoRngCore,
+        round: &Self::Round,
+    ) -> Result<<Self::Round as Round<Id>>::EchoBroadcast, LocalError> {
+        round.make_echo_broadcast(rng)
+    }
+
+    /// Called instead of [`Round::make_direct_message`].
+    ///
+    /// The default implementation calls [`Round::make_direct_message`].
+    #[allow(clippy::type_complexity)]
+    fn make_direct_message(
+        &self,
+        rng: &mut impl CryptoRngCore,
+        round: &Self::Round,
+        destination: &Id,
+    ) -> Result<
+        (
+            <Self::Round as Round<Id>>::DirectMessage,
+            <Self::Round as Round<Id>>::Artifact,
+        ),
+        LocalError,
+    > {
+        round.make_direct_message(rng, destination)
+    }
+
+    /// Called instead of [`Round::finalize`].
+    ///
+    /// The default implementation calls [`Round::finalize`].
+    fn extend_finalize(
+        &self,
+        rng: &mut impl CryptoRngCore,
+        round: Self::Round,
+        payloads: BTreeMap<Id, <Self::Round as Round<Id>>::Payload>,
+        artifacts: BTreeMap<Id, <Self::Round as Round<Id>>::Artifact>,
+    ) -> Result<FinalizeOutcome<Id, <Self::Round as Round<Id>>::Protocol>, LocalError> {
+        round.finalize(rng, payloads, artifacts)
+    }
+}
+
+#[derive_where::derive_where(Debug, Clone, Serialize, Deserialize)]
+struct ExtendedProtocolError<Id, Ext: Extension<Id>>(<Ext::Round as Round<Id>>::ProtocolError);
+
+impl<Id: PartyId, Ext: Extension<Id>> ProtocolError<Id> for ExtendedProtocolError<Id, Ext> {
+    type Round = ExtendedRound<Id, Ext>;
+    fn required_messages(&self, round_id: &RoundId) -> RequiredMessages {
+        self.0.required_messages(round_id)
+    }
+    fn verify_evidence(
+        &self,
+        round_id: &RoundId,
+        from: &Id,
+        shared_randomness: &[u8],
+        shared_data: &<<Self::Round as Round<Id>>::Protocol as Protocol<Id>>::SharedData,
+        messages: EvidenceMessages<'_, Id, Self::Round>,
+    ) -> Result<(), EvidenceError> {
+        let messages = messages.into_round::<Ext::Round>();
+        self.0
+            .verify_evidence(round_id, from, shared_randomness, shared_data, messages)
+    }
+    fn description(&self) -> String {
+        self.0.description()
+    }
+}
+
+#[allow(clippy::type_complexity)]
+#[derive(Debug)]
+struct ExtendedRound<Id, Ext: Extension<Id>> {
+    round: Ext::Round,
+    extension: Ext,
+    extensions: BTreeMap<TypeId, Box<dyn DynExtension<Id, <Ext::Round as Round<Id>>::Protocol>>>,
```

Comment:
Is my understanding of this correct? I don't think so, but perhaps you can elaborate.

```suggestion
    // Map between round types and  extensions. During protocol execution, this map is checked and if the current round has an extension defined, use it to extend the round. Otherwise fall through to the "normal" round.
    extensions: BTreeMap<TypeId, Box<dyn DynExtension<Id, <Ext::Round as Round<Id>>::Protocol>>>,
```

Reply:
> if the current round has an extension defined

More accurately, if the type of the current round has an extension defined.